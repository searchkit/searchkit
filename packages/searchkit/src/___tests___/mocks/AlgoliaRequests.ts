export const SimpleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequestOneFilter = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['actors:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const DisjunctiveExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'type',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const NestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const TwoFiltersNestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice', 'user.first:John']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFilterNestedQueryRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetFilters: [],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['user.price>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const FacetQueryExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetName: 'type',
      facetQuery: 'm',
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFiltersExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['imdbrating>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      clickAnalytics: false,
      facets: 'imdbrating',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 0,
      maxValuesPerFacet: 10,
      page: 0,
      query: ''
    }
  }
]

export const NestedFacetQueryExampleRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetName: 'user.first',
      facetQuery: 'har',
      facets: ['user.first', 'user.price'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]
