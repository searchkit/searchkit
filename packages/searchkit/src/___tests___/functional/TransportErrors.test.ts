import Client from '../../'
import { DisjunctiveExampleRequest } from '../mocks/AlgoliaRequests'
import type { AlgoliaMultipleQueriesQuery } from '@searchkit/api'
import { ESTransporter } from '../../'

import nock from 'nock'
import {
  APIKeyNotAuthorized,
  ElasticsearchError,
  InvalidRequest,
  KeyUnauthorizedForIndex
} from '../mocks/ErrorElasticsearchResponses'

describe('Transport Errors', () => {
  let consoleError: jest.SpyInstance
  beforeEach(() => {
    consoleError = jest.spyOn(console, 'error').mockImplementation(() => {})
  })

  const transporter = new ESTransporter(
    {
      host: 'http://localhost:9200',
      apiKey: 'apikey'
    },
    { debug: false }
  )

  it('401 error', async () => {
    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, APIKeyNotAuthorized)

    await expect(() => {
      return transporter.msearch([])
    }).rejects.toThrowErrorMatchingSnapshot()
    expect(consoleError).toBeCalled()
  })

  it('403 error', async () => {
    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, KeyUnauthorizedForIndex)

    await expect(() => {
      return transporter.msearch([])
    }).rejects.toThrowErrorMatchingSnapshot()
    expect(consoleError).toBeCalled()
  })

  it('400 error', async () => {
    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, InvalidRequest)

    await expect(() => {
      return transporter.msearch([])
    }).rejects.toThrowErrorMatchingSnapshot()
    expect(consoleError).toBeCalled()
  })

  it('500 error', async () => {
    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, ElasticsearchError)

    await expect(() => {
      return transporter.msearch([])
    }).rejects.toThrowErrorMatchingSnapshot()
    expect(consoleError).toBeCalled()
  })

  const client = new Client({
    connection: {
      host: 'http://localhost:9200',
      apiKey: 'apikey'
    },
    search_settings: {
      highlight_attributes: ['title', 'actors'],
      search_attributes: ['title', 'actors', 'query'],
      result_attributes: ['title', 'actors', 'query'],
      facet_attributes: ['type'],
      query_rules: []
    }
  })

  it('401 error', async () => {
    nock('http://localhost:9200')
      .post('/_msearch', (requestBody: any) => {
        return true
      })
      .reply(200, APIKeyNotAuthorized)

    const x = await client.handleInstantSearchRequests(
      DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[]
    )
    expect(x).toEqual({ results: [] })
    expect(consoleError).toBeCalled()
  })
})
