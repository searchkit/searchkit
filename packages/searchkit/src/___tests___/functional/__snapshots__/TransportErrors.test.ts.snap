// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transport Errors 400 error 1`] = `
"Elasticsearch Bad Request. 
          
          1. Check your query and make sure it is valid. 
          2. Check the field mapping. See documentation to make sure you are using text types for searching and keyword fields for faceting
          3. Turn on debug mode to see the Elasticsearch query and the error response.
          "
`;

exports[`Transport Errors 401 error 1`] = `"Cannot connect to Elasticsearch. Check your connection host and auth details (username/password or API Key required). You can also provide a custom Elasticsearch transporter to the API Client. See https://www.searchkit.co/docs/guides/setup-elasticsearch#connecting-with-usernamepassword for more details."`;

exports[`Transport Errors 403 error 1`] = `"Auth Error: You do not have permission to access this index. Check you are calling the right index (specified in frontend) and your API Key permissions has access to the index."`;

exports[`Transport Errors 500 error 1`] = `"Elasticsearch Internal Error: Check your elasticsearch instance to make sure it can recieve requests."`;
